rust/connlib/tunnel/src/peer.rs

```diff
@@ -129,10 +250,14 @@ impl ClientOnGateway {
         }
 
         let dst = packet.destination();
-        if self.resources.longest_match(dst).is_none() {
+        if !self
+            .filters
+            .longest_match(dst)
+            .is_some_and(|(_, filter)| filter.is_allowed(&packet.to_immutable()))
```

Comment:
Would it not combine the rules?

For example,

`10.0.0.0/24` -> Allow `tcp/80`
`10.0.0.1/32` -> Allow `tcp/443`

`10.0.0.1/32` would allow both `tcp/80` and `tcp/443`?

Reply:
We can do that but it requires some extra work, it would mean calculating what cidr are within another and adding those rules or iterating through all matches with each packet. 

If we want to do it I would go for the former although it's a bit harder to implement since this runs on the hot path.

Reply:
Could these be compiled into an efficient lookup structure whenever resources are updated or a DNS is resolved?

Reply:
yes, that's what I meant with:

> it would mean calculating what cidr are within another and adding those rules

I'll try to see how that looks

Reply:
This should do it https://github.com/firezone/firezone/pull/4779/commits/d9e894f55aa18d1886f5987f4059671177f6b416

Reply:
kk sounds good, yeah I suppose "compiling" the rules on the gateways make sense as opposed to taxing the portal with that task.

Reply:
Also, I think connlib have types already prepared to handle this

Reply:
There're some questions regarding resources added after connection is established, I will laid them in detail tomorrow and we can discuss in stand up

Reply:
Well, it's not only when resources are added after connection now that I think about it.

The clients request connections based for only a given ID, it doesn't know the filters for that given ID it only does based on IP and it uses the longest_match

Then the portal hands the resource based on the ID, so the gateway might never find out of the other more permissive resource.

So what we really need to do might require a bit much more work. It might be the client's responsability to pick the correct resource or the portal's.

My brain is not really working right now, will discuss more tomorrow

Reply:
This version will have to live with this problem as laid out here https://github.com/firezone/firezone/issues/2030#issuecomment-2078027977

But will be solved in a next PR with #4789

Reply:
Nice work on shipping a version that works with a plan to iterate on an improvement afterwards 💪